Get username from JWT token and remove tokens from command line args

MatthewPiatetsky opened the pull request:
Also only allow JWT tokens
@edx/ecommerce @rlucioni

MatthewPiatetsky commented on course_discovery/apps/course_metadata/data_loaders/api.py:
```diff
@@ -70,7 +71,11 @@ def ingest(self):
+            username = jwt.decode(self.access_token, verify=False).get('preferred_username', '')
+        except jwt.exceptions.DecodeError:
+            username = ''
+        response = self._make_request(initial_page, username)
```
If this is in fact only necessary for the CoursesApi it seems like we probably don't want to put the username variable in the abstract data loader, using self outside the init is a pylint error, which is I'm passing it here (though we could disable it)

clintonb replied:
Update `AbstractDataLoader` to accept `kwargs`. Pass the username as a `kwarg` and store it for this particular data loader. Move all of the decoding logic up to the management command level.

MatthewPiatetsky commented on course_discovery/apps/course_metadata/data_loaders/api.py:
```diff
@@ -70,7 +71,11 @@ def ingest(self):
         initial_page = 1
-        response = self._make_request(initial_page)
+        try:
+            username = jwt.decode(self.access_token, verify=False).get('preferred_username', '')
```
verify=False just pulls out the information without verifying the signature

clintonb replied:
Move this to the management command. Keep verification since we should always have a valid token.

MatthewPiatetsky replied:
isn't the secret for encoding the JWT token set in LMS or the extension that LMS uses?  
If catalog does not have that wouldn't I fail on signature verification?

MatthewPiatetsky commented:
did we want to remove the ability to pass token_type or remove the ability to pass a token command line altogether?

MatthewPiatetsky commented on course_discovery/apps/course_metadata/data_loaders/__init__.py:
```diff
@@ -23,10 +23,11 @@ class AbstractDataLoader(metaclass=abc.ABCMeta):
 
     PAGE_SIZE = 50
-    SUPPORTED_TOKEN_TYPES = ('bearer', 'jwt',)
+    SUPPORTED_TOKEN_TYPES = ('jwt',)
```
Should we remove this? Or keep if in the future we might use other types of tokens?

clintonb replied:
We should remove the ability to pass in a token altogether. We've never actually used it.

MatthewPiatetsky replied:
Ok. We actually say to use it here https://open-edx-course-catalog.readthedocs.io/en/latest/course_metadata.html#retrieving-course-metadata but I can update that

clintonb replied:
Correction: it is used by `ManagementViewSet.refresh_course_metadata()` which should also be removed.

clintonb commented on course_discovery/apps/course_metadata/data_loaders/api.py:
```diff
@@ -70,6 +70,7 @@ def ingest(self):
         logger.info('Refreshing Courses and CourseRuns from %s...', self.partner.courses_api_url)
 
         initial_page = 1
+
```
Revert.

clintonb commented on course_discovery/apps/course_metadata/management/commands/refresh_course_metadata.py:
```diff
@@ -116,6 +105,11 @@ def handle(self, *args, **options):
                     )
+                    try:
+                        username = jwt.decode(access_token, verify=False).get('preferred_username', '')
+                    except jwt.exceptions.DecodeError:
```
If decode fails, the entire command should fail. There is clearly something wrong with the token.

clintonb commented on course_discovery/apps/course_metadata/management/commands/refresh_course_metadata.py:
```diff
@@ -116,6 +105,11 @@ def handle(self, *args, **options):
                         token_type=token_type
                     )
+                    try:
+                        username = jwt.decode(access_token, verify=False).get('preferred_username', '')
```
If the token doesn't contain the `preferred_username` claim, the command should fail.

rlucioni commented on course_discovery/apps/course_metadata/management/commands/refresh_course_metadata.py:
```diff
@@ -98,27 +83,22 @@ def handle(self, *args, **options):
+                )
+                username = jwt.decode(access_token, verify=False)['preferred_username']
+                kwargs = {'username': username} if username else {}
+            except Exception:
```
Keep this exception handler confined to the token request. The decode should occur outside of it, in case, say, the preferred_username key is missing. If that were to occur, the message logged below would be confusing.

rlucioni commented on course_discovery/apps/course_metadata/management/commands/tests/test_refresh_course_metadata.py:
```diff
@@ -19,7 +19,7 @@
 from course_discovery.apps.course_metadata.tests.factories import CourseFactory
 
-ACCESS_TOKEN = 'secret'
+ACCESS_TOKEN = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiYWRtaW4iOnRydWUsInByZWZlcnJlZF91c2VybmFtZSI6ImpvaG4ifQ.SFJ4FVliPTP27EhxLHeTHDJJ6bTanoHFq-DWtMgUgPA'
```
Generate this in test code instead of hardcoding it.

MatthewPiatetsky replied:
why, isn't the actual token generation something that happens in the extension in LMS? There isn't actually code to test in this management command that generates a token

rlucioni replied:
In production-like deployments, JWTs come from our auth provider, which happens to be from the LMS. However, that doesn't preclude you from generating your own tokens for testing, which you can do with the `jwt` module. Doing that allows you to more clearly show what's being encoded in the token, and makes it easier to change in the future.

MatthewPiatetsky commented:
@clintonb @rlucioni I think I've addressed most of your feedback. 
As to the discussion on the other PR, I don't have a strong opinion on any approach, but it sounds like there are some differences of opinion. How do we think we will come to a decision?
(Note that most of the changes in this PR do not relate to pulling the username from the token, so if we decide not to do that, I can just remove those sections)

clintonb commented on course_discovery/apps/course_metadata/data_loaders/tests/mixins.py:
```diff
@@ -1,19 +1,18 @@
 @ddt.ddt
 class ApiClientTestMixin(object):
     @ddt.unpack
     @ddt.data(
```
`ddt` is no longer needed here.

clintonb commented on course_discovery/apps/course_metadata/management/commands/refresh_course_metadata.py:
```diff
@@ -98,27 +83,22 @@ def handle(self, *args, **options):
         if not partners:
             raise CommandError('No partners available!')
 
+        token_type = 'JWT'
```
This variable is not necessary.

MatthewPiatetsky replied:
It is used in a few places in this file. I could remove the need for specifying token type altogether, but imo it is helpful to explicitly clarify what the token_type is so people have an easier time understanding the auth.

clintonb commented on course_discovery/apps/course_metadata/management/commands/refresh_course_metadata.py:
```diff
@@ -98,27 +83,22 @@ def handle(self, *args, **options):
-                except Exception:
-                    logger.exception('No access token provided or acquired through client_credential flow.')
-                    raise
+            logger.info('No access token provided. Retrieving access token using client_credential flow...')
```
Update this: `Retrieving access token for partner [{partner.short_code}]...`

clintonb commented on course_discovery/apps/course_metadata/management/commands/tests/test_refresh_course_metadata.py:
```diff
@@ -38,9 +38,9 @@ def mock_apis(self):
+    def mock_access_token_api(self, requests_mock=None, callback=None):
         body = {
-            'access_token': ACCESS_TOKEN,
+            'access_token': str(jwt.encode({'preferred_username': 'bob'}, 'secret', algorithm='HS256'), 'utf-8'),
```
Store this in the constant.

clintonb commented on course_discovery/apps/course_metadata/management/commands/tests/test_refresh_course_metadata.py:
```diff
@@ -136,9 +136,16 @@ def test_refresh_course_metadata_with_invalid_partner_code(self):
-            command_args = ['--access_token=test-access-token']
-            call_command('refresh_course_metadata', *command_args)
+        with responses.RequestsMock() as rsps:
+            def raise_exception():
```
This seems dirty. Mock `EdxRestApiClient.get_oauth_access_token()` so that it raises an exception.

MatthewPiatetsky replied:
isn't that what I'm doing in mock_access_token_api?

MatthewPiatetsky replied:
got it, thanks!

MatthewPiatetsky commented:
@edx/ecommerce talked with @rlucioni @mikedikan during standup I believe the approach we settled on was finish this PR to get things working and simultaneously settle on an approach that everyone is happy with on the other discussion, and then create a dev story and pick it up (potentially next sprint)

rlucioni commented on course_discovery/apps/course_metadata/management/commands/refresh_course_metadata.py:
```diff
@@ -98,27 +83,22 @@ def handle(self, *args, **options):
+            except Exception:
+                logger.exception('No access token acquired through client_credential flow.')
+                raise
+            username = jwt.decode(access_token, verify=False)['preferred_username']
```
Why are you passing `verify=False` here? I understand that all we need to do here is read the contents of the token, but is it wise to do away with token verification in this context?

clintonb replied:
We do have the secret. Look at `settings.JWT_AUTH['JWT_SECRET']`.

rlucioni replied:
We currently use symmetric keys. The secrets are the same.

MatthewPiatetsky replied:
Locally, JWT_SECRET_KEY is not set for either, so it should fall back to SECRET_KEY
I checked the SECRET_KEY in my local catalog and local platform and they are not the same

rlucioni replied:
They are the same in working, production-like environments. We define values for JWT_SECRET_KEY in private settings specific to each developer.

MatthewPiatetsky replied:
so what should we do to make it work locally?  
And what if I want to point my local catalog to stage or prod LMS, if I don't want to run my local LMS  
Then I don't know what those secrets are, or if I am able to go to the environment and find out I will need to update my local settings everytime the secret changes  
This seems like a major complication to use this command locally

MatthewPiatetsky replied:
Since LMS is the one actually responsible for generating and verifying the token, it doesn't seem necessary to me to verify it on the catalog side. I would hope that if LMS cannot create a token, it would fail on token creation, rather than giving back some sort of broken token

rlucioni replied:
@MatthewPiatetsky anyone using/trusting the token is responsible for verifying its validity, not just the LMS. That said, in this context, I don't see any harm in extracting the username without checking the signature. I'm in favor of what you have here, unless @clintonb is opposed.

clintonb replied:
This is fine as-is.

rlucioni commented on course_discovery/apps/course_metadata/management/commands/tests/test_refresh_course_metadata.py:
```diff
@@ -19,8 +20,8 @@
 
-ACCESS_TOKEN = 'secret'
 JSON = 'application/json'
+ACCESS_TOKEN = str(jwt.encode({'preferred_username': 'bob'}, 'secret', algorithm='HS256'), 'utf-8')
```
The algorithm should already be specified in settings. Do you really need to provide it again here?

MatthewPiatetsky replied:
changed

rlucioni commented on docs/course_metadata.rst:
```diff
@@ -26,7 +26,7 @@ For example, if you use the JWT access token "secret-ecommerce-key" to authentic
 .. code-block:: bash
 
-    $ ./manage.py refresh_course_metadata --access_token secret-ecommerce-key --token_type jwt
+    $ ./manage.py refresh_course_metadata
```
Also change the line above this that refers to the JWT. Maybe make a note that the command retrieves it's own access token via the client credentials grant.

MatthewPiatetsky replied:
changed

clintonb commented on course_discovery/apps/course_metadata/management/commands/tests/test_refresh_course_metadata.py:
```diff
@@ -38,7 +39,7 @@ def mock_apis(self):
         self.mock_programs_api()
 
-    def mock_access_token_api(self, requests_mock=None):
+    def mock_access_token_api(self, requests_mock=None, callback=None):
```
The changes here are not necessary.

MatthewPiatetsky replied:
thanks, changed

clintonb commented on docs/course_metadata.rst:
```diff
@@ -21,12 +21,11 @@ Retrieving Course Metadata
 changed upstream then running it again should not change the database.
-
-For example, if you use the JWT access token "secret-ecommerce-key" to authenticate API calls, run the following:
+This command retrieves a JWT access token from the LMS instance that your OpenID Connect URL points to.
```
The command retrieves a JWT access token from the configured OpenID Connect provider.
